wpimath/src/main/native/include/frc/controller/ArmFeedforward.h

```diff
@@ -76,6 +80,53 @@ class WPILIB_DLLEXPORT ArmFeedforward {
            kV * velocity + kA * acceleration;
   }
 
+  /**
+   * Calculates the feedforward from the gains and setpoints.
+   *
+   * @param currentAngle        The current angle, in radians. This
+   * angle should be measured from the horizontal (i.e. if the provided angle is
+   * 0, the arm should be parallel to the floor). If your encoder does not
+   * follow this convention, an offset should be added.
+   * @param nextAngle        The next angle, in radians. This
+   * angle should be measured from the horizontal (i.e. if the provided angle is
+   * 0, the arm should be parallel to the floor). If your encoder does not
+   * follow this convention, an offset should be added.
+   * @param currentVelocity The current velocity setpoint, in radians per
+   *                        second.
+   * @param nextVelocity    The next velocity setpoint, in radians per second.
+   * @param dt              Time between velocity setpoints in seconds.
+   * @return The computed feedforward, in volts.
+   */
+  units::volt_t Calculate(units::unit_t<Angle> currentAngle,
+                          units::unit_t<Angle> nextAngle,
+                          units::unit_t<Velocity> currentVelocity,
+                          units::unit_t<Velocity> nextVelocity,
+                          units::second_t dt) const {
+    Vectord<2> r{currentAngle.value(), currentVelocity.value()};
+    Vectord<2> nextR{nextAngle.value(), nextVelocity.value()};
+
+    sleipnir::OptimizationProblem problem;
+    auto u_k = problem.DecisionVariable();
+
+    Matrixd<2, 2> A{{0.0, 1.0}, {0.0, -kV.value() / kA.value()}};
+    Matrixd<2, 1> B{0.0, 1.0 / kA.value()};
+    Matrixd<2, 1> C{0.0, (-kS / kA) * wpi::sgn(currentVelocity.value()) -
+                             (kG / kA) * units::math::cos(currentAngle)};
+
+    auto f = [&](const Vectord<2>& x, const Vectord<1>& u) -> Vectord<2> {
+      return A * x + B * u + C;
+    };
+
+    auto residual =
+        (nextR - RK4(f, r, Vectord<1>{u_k.Value()}, dt)).transpose() *
+        (nextR - RK4(f, r, Vectord<1>{u_k.Value()}, dt));
```

Comment:
How computationally expensive is it to calculate `nextR - RK4(f, r, Vectord<1>{u_k.Value()}, dt)`? Would it make sense to store it into a variable and compute once?

Reply:
All of this is based on the Sleipnir implementation to WPILIB provided by @calcmogul.  
Since this is an optimization problem this cannot be stored in a variable.  The variable u_k is the optimization variable and is what is solved for.

Reply:
Here is a [link ](https://github.com/calcmogul/allwpilib/tree/upstream-utils-add-sleipnir)to his working branch that I based this PR off of including the details listed in [#5192 ](https://github.com/wpilibsuite/allwpilib/issues/5192)

Reply:
As to how expensive it is to calculate this.  I cannot speak to that directly, but would love feedback on determining that.

Reply:
I think caching this is a good decision